Demo of the full Greentea protocol flow on Unix PTY

LDong-Arm opened the pull request:
Add a demo which creates a PTY (pseudo-terminals) device node which the host (htrun) can talk to like a serial device. To run it, two shells are required: one for the device running greentea-client, one for the host.

Notes:
* Given that htrun currently doesn't provide a generic way (e.g. stdio) to interact with another process running greentea-client, PTY is the easiest way to demo the full Greentea protocol without requiring an embedded device or Mbed OS.
* PTY requires a Unix-like OS. Windows has a different API to access terminals, so this example is excluded on Windows and only the client library itself is built.

Also remove the CMake cache variable `BUILD_EXAMPLES`. Its intention was to offer an option to not build the examples when greentea-client is embedded in another project, but now we already have the check `if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)` which does it more elegantly.

LDong-Arm committed:
Full device (greentea-client) & host (htrun) demo

Add a demo which creates a PTY (pseudo-terminals) device node which
the host (htrun) can talk to like a serial device. To run it, two
shells are required: one for the device running greentea-client, one
for the host.

Notes:
* Given that htrun currently doesn't provide a generic way
(e.g. stdio) to interact with another process running greentea-client,
PTY is the easiest way to demo the full Greentea protocol without
requiring an embedded device or Mbed OS.
* PTY requires a Unix-like OS. Windows has a different API to access
terminals, so this example is excluded on Windows.

LDong-Arm commented on examples/pty/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# demo does not support yet.
+set(SUPPORTED_SYSTEMS Darwin Linux)
+if(NOT CMAKE_SYSTEM_NAME IN_LIST SUPPORTED_SYSTEMS)
+    message(FATAL_ERROR "The PTY example supports macOS and Linux only.")
```
Maybe it should be a warning instead, because this directory is added by the top-level CMakeLists.txt and we shouldn't prevent Windows users to building the library itself.

rwalton-arm replied:
That's probably sensible. A related issue I have is that BUILD_EXAMPLES is normally ON, which means we always build examples with `make all`. I think the user should have to specify that they want to build the examples rather than building them by default.

LDong-Arm replied:
I thought we wanted to avoid `EXCLUDE_FROM_ALL`, so the examples are built by default when building the current project as the main project (#26). At least I'll remove the fatal error.

rwalton-arm replied:
OK, fair enough. We can just remove the error for now.

However, I don't really understand why we wanted to avoid `EXCLUDE_FROM_ALL` in this case, as `EXCLUDE_FROM_ALL` is basically intended for optional examples. That use case is even mentioned in the [cmake docs](https://cmake.org/cmake/help/latest/command/add_subdirectory.html). I know we want to change the way we use `EXCLUDE_FROM_ALL` in Mbed OS, because we are forced to abuse it due to the way the Mbed OS source tree is structured. Anyway, we can discuss that more on Monday, as it's a design discussion which is out of scope for this PR.

LDong-Arm commented:
Pseudoconsole on Windows:
https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session

Patater commented on examples/pty/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# demo does not support yet.
+set(SUPPORTED_SYSTEMS Darwin Linux)
+if(NOT CMAKE_SYSTEM_NAME IN_LIST SUPPORTED_SYSTEMS)
+    message(WARNING "Skipping the PTY example which supports macOS and Linux only.")
```
This says "skipping", but how is the skipping actually happening?

LDong-Arm replied:
Good catch, now fixed

Patater reviewed (changes requested):
"This replaces the file-based custom_io example, as there is no point
maintaining two examples to demonstrate how to override the default
IO."

The PTY example is more complex, so less good of an example for custom_io. I'd recommend keeping both for at least that reason.

LDong-Arm commented:
> "This replaces the file-based custom_io example, as there is no point
> maintaining two examples to demonstrate how to override the default
> IO."
> 
> The PTY example is more complex, so less good of an example for custom_io. I'd recommend keeping both for at least that reason.

Thanks for the suggestions, I've brought back custom_io and updated the README.

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -88,12 +88,11 @@ endif()
-set(BUILD_EXAMPLES ON CACHE BOOL "Enable building the examples")
-
-if((CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME) AND BUILD_EXAMPLES)
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
I removed `BUILD_EXAMPLES`, because the intention of was to offer an option to not build the examples when greentea-client is embedded in another project. But now we already have the check `if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)` which does it more elegantly.

LDong-Arm replied:
Just realised `BUILD_EXAMPLES` and `if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)` were introduced at the same time (#26), so I might have misunderstood the intention?

Patater replied:
The option is there so you can exclude building examples if you'd like to, even when not embedded in another project.

LDong-Arm replied:
But `-DBUILD_EXAMPLES=OFF` doesn't look simpler than `--target client`?

rwalton-arm replied:
`--target client` is an implementation detail, `BUILD_EXAMPLES` is part of a "public interface". My PR actually changes `BUILD_EXAMPLES` to be a CMake option, because that's the semantically correct way of exposing options to the user, rather than using an internal cache variable.

LDong-Arm replied:
I'm going to revert this part to what it was before as it's not related to this PR. Will let your PR handle it or have discussion there.

rwalton-arm replied:
I'm still not sure whether keeping the option or using EXCLUDE_FROM_ALL and making the example targets part of the stable public interface is the better approach, though. 

Maybe it doesn't really matter too much which approach we take here, as the vast majority of users will be depending on this library in their test applications rather than building it alone just to run examples. So, maybe always building the examples if this is the current project is actually fine.

LDong-Arm replied:
Yea I agree it doesn't really matter here.

Bundling example(s) in a library's repository is quite common, so hopefully we'll reach a conclusion and follow that in the future. I've reverted this part for now.

rwalton-arm commented:
> Pseudoconsole on Windows:
> https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session

Should we also add Windows support to this demo, or is that not feasible?

LDong-Arm commented:
> Should we also add Windows support to this demo, or is that not feasible?

@rwalton-arm I don't have a Windows environment to develop and test it, but if anyones wants to add support for Windows, they can raise another PR.

The long-term solution @Patater suggests is adding an option in htrun to use stdio or files: https://github.com/ARMmbed/greentea/issues/305 which doesn't quite depend on the OS.

rwalton-arm commented:
> > Should we also add Windows support to this demo, or is that not feasible?
> 
> @rwalton-arm I don't have a Windows environment to develop and test it, but if anyones wants to add support for Windows, they can raise another PR.
> 
> The long-term solution @Patater suggests is adding an option in htrun to use stdio or files: [ARMmbed/greentea#305](https://github.com/ARMmbed/greentea/issues/305) which doesn't quite depend on the OS.

Yes, an "attach to process" feature, where htrun would read the process's stdout, would be useful. Sounds like the right solution to me.

Patater reviewed (changes requested):
The renaming of the stdio example should be its own commit.

The pty example commit sill says "This commit also removes the CMake cache variable `BUILD_EXAMPLES`.", but that isn't the case anymore and should be removed from the commit message.

LDong-Arm committed:
Rename the examples

Previously we used the same name `greentea-client-example` for both
the stdio and custom_io examples, and relied on the switch
`GREENTEA_CLIENT_STDIO` to control which one gets built. But as
different examples, each one should have its own precise name for
clarity and to allow multiple examples to be built in the future.

LDong-Arm commented:
> The renaming of the stdio example should be its own commit.
> 
> The pty example commit sill says "This commit also removes the CMake cache variable `BUILD_EXAMPLES`.", but that isn't the case anymore and should be removed from the commit message.

@Patater Both are resolved now, thanks
